execute/tokendata/observer.go

```diff
@@ -2,42 +2,145 @@ package tokendata
+	// If TokenDataObserver doesn't support the token it must return exectypes.NotSupportedTokenData.
+	// This way higher layer will know how to orchestrate Observers and which observations are relevant.
+	//
+	// TokenDataObserver must return data for every token that is present in the exectypes.MessageObservations.
+	// Meaning that exectypes.TokenDataObservations must reflect the structure of exectypes.MessageObservations
+	// (the same number of objects and tokens).
+	// * if token is supported and requires offchain processing, TokenDataObserver must initialize the token according
+	// 	 to its logic. Depending on the result it must initialize with either exectypes.NewSuccessTokenData or
+	//	 exectypes.NewErrorTokenData
+	// * if token is not supported, TokenDataObserver must set the token data to exectypes.NotSupportedTokenData
 	Observe(
 		ctx context.Context,
 		observations exectypes.MessageObservations,
 	) (exectypes.TokenDataObservations, error)
+
+	// IsTokenSupported returns true if the token is supported by the observer.
+	IsTokenSupported(sourceChain cciptypes.ChainSelector, msgToken cciptypes.RampTokenAmount) bool
 }
 
 // CompositeTokenDataObserver is a TokenDataObserver that combines multiple TokenDataObserver behind the same interface.
 // Goal of that is to support multiple token observers supporting different tokens (e.g. CCTP, MyFancyToken etc)
 type CompositeTokenDataObserver struct {
-	Observers []TokenDataObserver
+	observers []TokenDataObserver
+}
+
+// nolint early-return
+func NewConfigBasedCompositeObservers(config []pluginconfig.TokenDataObserverConfig) (*CompositeTokenDataObserver, error) {
+	observers := make([]TokenDataObserver, len(config))
+	for i, c := range config {
+		if c.USDCCCTPObserverConfig != nil {
+			observers[i] = usdc.NewUSDCCCTPTokenDataObserver(*c.USDCCCTPObserverConfig)
+		} else {
+			return nil, errors.New("unsupported token data observer")
+		}
+	}
+	return NewCompositeObservers(observers...), nil
+}
+
+func NewCompositeObservers(observers ...TokenDataObserver) *CompositeTokenDataObserver {
+	return &CompositeTokenDataObserver{observers: observers}
 }
 
+// Observe start with stubbing exectypes.TokenDataObservations with empty data based on the supported tokens.
+// Then it iterates over all observers and merges token data returned from them into the final result.
 func (t *CompositeTokenDataObserver) Observe(
 	ctx context.Context,
-	messages exectypes.MessageObservations,
+	msgObservations exectypes.MessageObservations,
 ) (exectypes.TokenDataObservations, error) {
-	var tokenObservation exectypes.TokenDataObservations
-	for _, ob := range t.Observers {
-		tokenData, err := ob.Observe(ctx, messages)
+	tokenDataObservations := t.initTokenDataObservations(msgObservations)
+
+	for _, ob := range t.observers {
+		tokenData, err := ob.Observe(ctx, msgObservations)
+		if err != nil {
+			return nil, err
```

Comment:
Why return if one observer has an error? Others can complete the operation successfully.

Reply:
Maybe also check IsTokenSupported before the Observe call?

Reply:
> Maybe also check IsTokenSupported before the Observe call?

It doesn't work this way, every Observer gets all the data, if it doesn't support any tokens it should just bypass and return empty data. Also `IsTokenSupported` works on a token level, whereas we do batching here and pass everything. Therefore underlying implementation can batch requests more efficiently

Reply:
> Why return if one observer has an error? Others can complete the operation successfully.

Yeah, I guess you are right, no need to return an error here. I think it's not that relevant at this stage as we have only a single observer, but let me change that and create a test for that